internal/suggest/errorCodes.go

```diff
@@ -0,0 +1,320 @@
+	schemaName, responseName string
+}
+type errorGroupSlice []errorGroup
+
+func initErrorGroups() errorGroupSlice {
+	return errorGroupSlice{{
+		name:         "BadRequest",
+		codes:        []string{"400", "413", "414", "415", "422", "431", "510"},
+		description:  "A collection of codes that generally means the end user got something wrong in making the request",
+		schemaName:   "BadRequestError",
+		responseName: "BadRequest",
+	},
+		{
+			name:         "Unauthorized",
+			codes:        []string{"401", "403", "407", "511"},
+			description:  "A collection of codes that generally means the client was not authenticated correctly for the request they want to make",
+			schemaName:   "UnauthorizedError",
+			responseName: "Unauthorized",
+		},
+		{
+			name:         "NotFound",
+			codes:        []string{"404", "501", "505"},
+			description:  "Status codes relating to the resource/entity they are requesting not being found or endpoints/routes not existing",
+			schemaName:   "NotFoundError",
+			responseName: "NotFound",
+		},
+		{
+			name:         "RateLimited",
+			codes:        []string{"429"},
+			description:  "Status codes relating to the client being rate limited by the server",
+			schemaName:   "RateLimitedError",
+			responseName: "RateLimited",
+		},
+		{
+			name:         "InternalServerError",
+			codes:        []string{"500", "502", "503", "506", "507", "508"},
+			description:  "A collection of status codes that generally mean the server failed in an unexpected way",
+			schemaName:   "InternalServerError",
+			responseName: "InternalServerError",
+		},
+		{
+			name:         "Timeout",
+			codes:        []string{"408", "504"},
+			description:  "Timeouts occurred with the request",
+			schemaName:   "TimeoutError",
+			responseName: "Timeout",
+		},
+	}
+}
+
+func (e errorGroupSlice) FindCode(code string) errorGroup {
+	for _, group := range e {
+		if slices.Contains(group.codes, code) {
+			return group
+		}
+	}
+	return errorGroup{}
+}
+
+func (e errorGroupSlice) AllCodes() []string {
+	var codes []string
+	for _, group := range e {
+		codes = append(codes, group.codes...)
+	}
+	return codes
+}
+
+func (e errorGroupSlice) DeduplicateComponentNames(document v3.Document) {
+	var schemaNames []string
+	for s := range openapi.IterateSchemas(document) {
+		schemaNames = append(schemaNames, s.Name)
+	}
+
+	var responseNames []string
+	for r := range openapi.IterateResponses(document) {
+		responseNames = append(responseNames, r.Name)
+	}
+
+	for i, group := range e {
+		e[i].responseName = findUnusedName(group.responseName, responseNames)
+		e[i].schemaName = findUnusedName(group.schemaName, schemaNames)
+	}
+}
+
+func BuildErrorCodesOverlay(ctx context.Context, document v3.Document) (*overlay.Overlay, error) {
+	groups := initErrorGroups()
+	groups.DeduplicateComponentNames(document)
+
+	builder := builder{document: document, errorGroups: groups}
+	return builder.Build()
+}
+
+type builder struct {
```

Comment:
Do you think it might be worth having a common interface for implementing suggest types here?

Reply:
e.g at its most simplest:

```
type Suggester interface {
	Suggest(document v3.Document) (*overlay.Overlay, error)
}
```

Reply:
Reason being is that I am / going to be implementing a similar suggest type for pagination here

Reply:
I do think that could be valuable! I love an interface. At the moment i don't have a use case for it, but when one appears I think the refactor would be worthwhile

Reply:
I'm currently on the pagination stuff on a different branch and have implemented something like this so I think it would be best if we aligned on this early. I think also it sets up future suggest types for success and quick implementation

Reply:
FWIW this `builder` thing is just an internal util to hold some state rather than passing it around

Reply:
Whatever interface you want to introduce im happy to adjust this to implement